servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,257 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.encoding.netty;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.buffer.api.CompositeBuffer;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Publisher;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.MessageToByteEncoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.netty.util.internal.PlatformDependent.throwException;
+import static io.servicetalk.buffer.api.EmptyBuffer.EMPTY_BUFFER;
+import static io.servicetalk.buffer.api.ReadOnlyBufferAllocators.DEFAULT_RO_ALLOCATOR;
+import static io.servicetalk.buffer.netty.BufferUtils.newBufferFrom;
+import static io.servicetalk.buffer.netty.BufferUtils.toByteBuf;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static java.util.Objects.requireNonNull;
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderFn;
+    private final Supplier<ByteToMessageDecoder> decoderFn;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderFn,
+                             final Supplier<ByteToMessageDecoder> decoderFn) {
+        super(name);
+        this.encoderFn = encoderFn;
+        this.decoderFn = decoderFn;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length,
+                         final BufferAllocator allocator) {
+        requireNonNull(src);
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderFn.get();
+        final EmbeddedChannel channel = new EmbeddedChannel(encoder);
```

Comment:
By default, the `EmbeddedChannel` is created with netty's `ByteBufAllocator` and `RecvByteBufAllocator`. Both will use pooled memory that we may never release. We should override both to use  the passed allocator which have to be a `ServiceTalkBufferAllocator`. You can do that using `channel.config()`.